Examples/Io/EDM4hep/src/EDM4hepReader.cpp

```diff
@@ -0,0 +1,467 @@
+    }
+    return result;
+  };
+
+  const std::size_t gen = parentId.generation();
+  maxGen = std::max(maxGen, gen);
+
+  ACTS_VERBOSE(indent(gen) << "  - processing daughters for input particle "
+                           << inParticle.id());
+  ACTS_VERBOSE(indent(gen) << "    -> found " << inParticle.daughters_size()
+                           << " daughter(s)");
+
+  bool parentDecayed =
+      std::any_of(inParticle.daughters_begin(), inParticle.daughters_end(),
+                  [](const edm4hep::MCParticle& daughter) {
+                    return !daughter.vertexIsNotEndpointOfParent();
+                  });
+  std::size_t secondaryVertex = 0;
+  if (parentDecayed) {
+    ACTS_VERBOSE(indent(gen) << "    -> parent decays");
+    secondaryVertex = ++nSecondaryVertices;
+  }
+
+  std::size_t parentIndex = particles.size() - 1;
+
+  std::size_t nParticles = 0;
+  for (const auto& daughter : inParticle.getDaughters()) {
+    SimParticle particle = EDM4hepUtil::readParticle(daughter);
+
+    auto pid = parentId.makeDescendant(nParticles++);
+    if (daughter.vertexIsNotEndpointOfParent()) {
+      // incoming particle survived, interaction via descendant
+    } else {
+      // incoming particle decayed
+      pid = pid.setVertexSecondary(secondaryVertex);
+    }
+    particle.setParticleId(pid);
+
+    ACTS_VERBOSE(indent(particle.particleId().generation())
+                 << "+ add particle " << particle);
+    ACTS_VERBOSE(indent(particle.particleId().generation())
+                 << "  - generation: " << particle.particleId().generation());
+    ACTS_VERBOSE(indent(particle.particleId().generation())
+                 << "  - at " << particle.position().transpose());
+    ACTS_VERBOSE(indent(particle.particleId().generation())
+                 << "     - createdInSim: "
+                 << daughter.isCreatedInSimulation());
+    ACTS_VERBOSE(indent(particle.particleId().generation())
+                 << "     - vertexIsNotEndpointOfParent: "
+                 << daughter.vertexIsNotEndpointOfParent());
+    ACTS_VERBOSE(indent(particle.particleId().generation())
+                 << "     - isStopped: " << daughter.isStopped());
+    ACTS_VERBOSE(indent(particle.particleId().generation())
+                 << "     - endpoint: " << daughter.getEndpoint().x << ", "
+                 << daughter.getEndpoint().y << ", "
+                 << daughter.getEndpoint().z);
+
+    particles.push_back(std::move(particle));
+    particleMap[daughter.getObjectID().index] = particles.size() - 1;
+    parentRelationship[particles.size() - 1] = parentIndex;
+    processChildren(daughter, pid, particles, parentRelationship, particleMap,
+                    nSecondaryVertices, maxGen);
+  }
+}
+
+void EDM4hepReader::setSubParticleIds(
+    SimParticleContainer::sequence_type::iterator begin,
+    SimParticleContainer::sequence_type::iterator end) {
+  std::vector<std::size_t> numByGeneration;
+  numByGeneration.reserve(10);
```

Comment:
where does 10 come from?

Reply:
Arbitrary guess what the upper limit of number of generations is.